README
File-BaseDir version 0.1
=========================

This module can be used to find directories and files as specified
by the XDG Base Directory Specification.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules which can be obtained from
the CPAN <http://cpan.org> if they are not allready installed on
your system :

  Carp
  Exporter
  File::Spec

COPYRIGHT AND LICENCE

Copyright (c) 2003 Jaap G Karssenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.09      2021-07-08 14:04:14 -0600
 - Tests can now be run in parallel or isolation safely (rt#199256)
 - Fallback on getpwuid if environment variable not set (rt#127138)
 - Test suite works with or without environment variables (rt#127138)

0.08 13 Mar 2018
    Fix for failing test in 03_userdirs.t, thanks to Pierre Labastie

0.07 21 Apr 2015
    Use temporary files for 03_userdirs.t test, fix for RT bug 103805

0.06 6 Apr 2015
    Added prerequsite module build_requires section of Build.PL

0.05 4 Apr 2015
    Added prerequsitie module to Build.PL

0.04 3 Apr 2015
    Fix for RT bug #41744 . Remove warning message when $ENV{HOME} is not defined
    Fix for RT bugs # 88502 and #54929, typos corrected
	Fix for RT bug #41744 . Remove warning message when $ENV{HOME} is not defined
    Fix for RT bugs #50343, added more functionality
	Fixed some OS depedendancies in test suite.
	

0.03  Wed Oct 24 2007
	- Added lookup methods for files and dirs:
		data_home, data_files, data_dirs,
		config_home, config_files, config_dirs & cache_home;
		suggested by Matisse Enzer
	- Deprecated xdg_data_files and xdg_config_files;
		suggested by Michael Schwern
	- Added a bit of Win32 support
	- Extended tests to full coverage

0.02  Thu Sep  4 2003
        - forgot to remove h2xs default "use 5.008" tag :(
	emergency release it is ...

0.01  Wed Sep  3 18:33:59 2003
	- original version; created by h2xs 1.22 with options
		-AX -n File::BaseDir
	- Added methods for all environment variables defined by the spec



lib/File/UserDirs.pm
package File::UserDirs;

use strict;
use warnings;
use IPC::System::Simple qw(capturex);
use Exporter 5.57 qw( import );

# ABSTRACT: Find extra media and documents directories
# VERSION

our %EXPORT_TAGS = (
    all => [
        qw(xdg_desktop_dir xdg_documents_dir xdg_download_dir xdg_music_dir
        xdg_pictures_dir xdg_publicshare_dir xdg_templates_dir xdg_videos_dir)
    ]);
our @EXPORT_OK = @{$EXPORT_TAGS{all}};


sub _xdg_user_dir {
    my ($purpose) = @_;
    my $dir = capturex 'xdg-user-dir', $purpose;
    chomp $dir;
    return $dir;
}

sub xdg_desktop_dir     {return _xdg_user_dir 'DESKTOP';}
sub xdg_documents_dir   {return _xdg_user_dir 'DOCUMENTS';}
sub xdg_download_dir    {return _xdg_user_dir 'DOWNLOAD';}
sub xdg_music_dir       {return _xdg_user_dir 'MUSIC';}
sub xdg_pictures_dir    {return _xdg_user_dir 'PICTURES';}
sub xdg_publicshare_dir {return _xdg_user_dir 'PUBLICSHARE';}
sub xdg_templates_dir   {return _xdg_user_dir 'TEMPLATES';}
sub xdg_videos_dir      {return _xdg_user_dir 'VIDEOS';}

1;

__END__

=head1 SYNOPSIS

 use File::UserDirs qw(:all);
 print xdg_desktop_dir; # e.g. /home/user/Desktop

=head1 DESCRIPTION

This module can be used to find directories as informally specified
by the Freedesktop.org xdg-user-dirs software. This
gives a mechanism to locate extra directories for media and documents files.

=head1 FUNCTIONS

May be exported on request.
Also the group C<:all> is defined which exports all methods.

=head2 xdg_desktop_dir

 my $dir = xdg_desktop_dir;

Returns the desktop directory. Unless changed by the user,
this is the directory F<Desktop> in the home directory.

=head2 xdg_documents_dir

 my $dir = xdg_documents_dir;

Returns the documents directory. Unless changed by the user,
this is the home directory.

=head2 xdg_download_dir

 my $dir = xdg_download_dir;

Returns the download directory. Unless changed by the user,
this is the home directory.

=head2 xdg_music_dir

 my $dir = xdg_music_dir;

Returns the music directory. Unless changed by the user,
this is the home directory.

=head2 xdg_pictures_dir

 my $dir = xdg_pictures_dir;

Returns the pictures directory. Unless changed by the user,
this is the home directory.

=head2 xdg_publicshare_dir

 my $dir = xdg_publicshare_dir;

Returns the public share directory. Unless changed by the user,
this is the home directory.

=head2 xdg_templates_dir

 my $dir = xdg_templates_dir;

Returns the templates directory. Unless changed by the user,
this is the home directory.

=head2 xdg_videos_dir

 my $dir = xdg_videos_dir;

Returns the videos directory. Unless changed by the user,
this is the home directory.

=head1 DIAGNOSTICS

=over

=item C<"xdg-user-dir" failed to start: %s>

The executable C<xdg-user-dir> could not be run, most likely because it was not
installed. See L</"DEPENDENCIES">.

=back

=head1 CONFIGURATION AND ENVIRONMENT

The location of the directories can be specified by the user in the file
F<$XDG_CONFIG_HOME/user-dirs.dirs>. It is a shell file setting a number of
environment variables. To find the exact pathname from Perl, run:

 use File::BaseDir qw(config_home);
 print config_home('user-dirs.dirs');

=head2 Example customised F<user-dirs.dirs>

 XDG_DESKTOP_DIR="$HOME/Workspace"
 XDG_DOCUMENTS_DIR="$HOME/Files"
 XDG_DOWNLOAD_DIR="$HOME/Files/Downloads"
 XDG_MUSIC_DIR="$HOME/Files/Audio"
 XDG_PICTURES_DIR="$HOME/Files/Images"
 XDG_PUBLICSHARE_DIR="$HOME/public_html"
 XDG_TEMPLATES_DIR="$HOME/Files/Document templates"
 XDG_VIDEOS_DIR="$HOME/Files/Video"

=head1 DEPENDENCIES

This module requires the executable F<xdg-user-dir> from the package
C<xdg-user-dirs>. Source code is available from
L<http://cgit.freedesktop.org/xdg/xdg-user-dirs/>.

=cut


lib/File/IconTheme.pm
package File::IconTheme;

use strict;
use warnings;
use File::BaseDir qw(data_dirs);
use File::Spec;
use Exporter 5.57 qw( import );

# ABSTRACT: Find icon directories
# VERSION

our @EXPORT_OK = qw(xdg_icon_theme_search_dirs);

sub xdg_icon_theme_search_dirs {

    return grep {-d $_ && -r $_}
        File::Spec->catfile(File::BaseDir->_home, '.icons'),
        data_dirs('icons'),
        '/usr/share/pixmaps';
}

1;

__END__

=head1 SYNOPSIS

 use File::IconTheme qw(xdg_icon_theme_search_dirs);
 print join "\n", xdg_icon_theme_search_dirs;

=head1 DESCRIPTION

This module can be used to find directories as specified
by the Freedesktop.org Icon Theme Specification. Currently only a tiny
(but most useful) part of the specification is implemented.

In case you want to B<store> an icon theme, use the directory returned by:

 use File::BaseDir qw(data_dirs);
 print scalar data_dirs('icons');

=head1 FUNCTIONS

Can be exported on request.

=head2 xdg_icon_theme_search_dirs

 my @dirs = xdg_icon_theme_search_dir;

Returns a list of the base directories of icon themes.

=head1 CONFIGURATION AND ENVIRONMENT

C<$XDG_DATA_HOME>, C<$XDG_DATA_DIRS>

=head1 SEE ALSO

L<http://standards.freedesktop.org/icon-theme-spec/>

=cut


lib/File/BaseDir.pm
package File::BaseDir;

use strict;
use warnings;
use Exporter 5.57 qw( import );
use File::Spec;
use Config;

# ABSTRACT: Use the Freedesktop.org base directory specification
# VERSION

our %EXPORT_TAGS = (
  vars => [ qw(
    xdg_data_home xdg_data_dirs
    xdg_config_home xdg_config_dirs
    xdg_cache_home
  ) ],
  lookup => [ qw(
    data_home data_dirs data_files
    config_home config_dirs config_files
    cache_home
  ) ],
);
our @EXPORT_OK = (
  qw(xdg_data_files xdg_config_files),
  map @$_, values %EXPORT_TAGS
);

if($^O eq 'MSWin32')
{
  *_rootdir = sub { 'C:\\' };
  *_home    = sub { $ENV{USERPROFILE} || $ENV{HOMEDRIVE}.$ENV{HOMEPATH} || 'C:\\' };
}
else
{
  *_rootdir = sub { File::Spec->rootdir };
  *_home    = sub { $ENV{HOME} || eval { [getpwuid($>)]->[7] } || File::Spec->rootdir };
}

# OO method
sub new { bless \$VERSION, shift } # what else is there to bless ?

# Variable methods
sub xdg_data_home { $ENV{XDG_DATA_HOME} || File::Spec->catdir(_home(), qw/.local share/) }

sub xdg_data_dirs {
  ( $ENV{XDG_DATA_DIRS}
    ? _adapt($ENV{XDG_DATA_DIRS})
    : (File::Spec->catdir(_rootdir(), qw/usr local share/), File::Spec->catdir(_rootdir(), qw/usr share/))
  )
}

sub xdg_config_home {$ENV{XDG_CONFIG_HOME} || File::Spec->catdir(_home(), '.config') }

sub xdg_config_dirs {
  ( $ENV{XDG_CONFIG_DIRS}
    ? _adapt($ENV{XDG_CONFIG_DIRS})
    : File::Spec->catdir(_rootdir(), qw/etc xdg/)
  )
}

sub xdg_cache_home { $ENV{XDG_CACHE_HOME} || File::Spec->catdir(_home(), '.cache') }

sub _adapt {
  map { File::Spec->catdir( split(/\//, $_) ) } split /\Q$Config{path_sep}\E/, shift;
    # ':' defined in the spec, but ';' is standard on win32
}

# Lookup methods
sub data_home { _catfile(xdg_data_home, @_) }

sub data_dirs { _find_files(\&_dir, \@_, xdg_data_home, xdg_data_dirs) }

sub data_files { _find_files(\&_file, \@_, xdg_data_home, xdg_data_dirs) }

sub xdg_data_files { my @dirs = data_files(@_); return @dirs }

sub config_home { _catfile(xdg_config_home, @_) }

sub config_dirs { _find_files(\&_dir, \@_, xdg_config_home, xdg_config_dirs) }

sub config_files { _find_files(\&_file, \@_, xdg_config_home, xdg_config_dirs) }

sub xdg_config_files { my @dirs = config_files(@_); return @dirs }

sub cache_home { _catfile(xdg_cache_home, @_) }

sub _catfile {
  my $dir = shift;
  shift if ref $_[0] or $_[0] =~ /::/; # OO call
  return File::Spec->catfile($dir, @_);
}

sub _find_files {
  my $type = shift;
  my $file = shift;
  shift @$file if ref $$file[0] or $$file[0] =~ /::/; # OO call
  #warn "Looking for: @$file\n         in: @_\n";
  if (wantarray) {  ## no critic (Community::Wantarray)
    return grep { &$type( $_ ) && -r $_ }
           map  { File::Spec->catfile($_, @$file) } @_;
  }
  else { # prevent unnecessary stats by returning early
    for (@_) {
      my $path = File::Spec->catfile($_, @$file);
      return $path if &$type($path) && -r $path;
    }
  }
  return ();
}

sub _dir { -d $_[0] }

sub _file { -f $_[0] }

1;

__END__

=head1 SYNOPSIS

 use File::BaseDir qw/xdg_data_files/;
 for ( xdg_data_files('mime/globs') ) {
   # do something
 }

=head1 DESCRIPTION

This module can be used to find directories and files as specified
by the Freedesktop.org Base Directory Specification. This specifications
gives a mechanism to locate directories for configuration, application data
and cache data. It is suggested that desktop applications for e.g. the
GNOME, KDE or Xfce platforms follow this layout. However, the same layout can
just as well be used for non-GUI applications.

This module forked from L<File::MimeInfo>.

This module follows version 0.6 of BaseDir specification.

=head1 CONSTRUCTOR

=head2 new

 my $bd = File::BaseDir->new;

Simple constructor to allow calling functions as object oriented methods.

=head1 FUNCTIONS

None of these are exported by default, but all functions can be exported
by request.  Also the groups C<:lookup> and C<:vars> are defined.  The
C<:vars> group contains all the routines with a C<xdg_> prefix. The
C<:lookup> group contains the routines to locate files and directories.

=head2 data_home

 my $path = data_home(@path);
 my $path = $bd->data_home(@path);

Takes a list of file path elements and returns a new path by appending
them to the data home directory. The new path does not need to exist.
Use this when writing user specific application data.

Example:

 # data_home is: /home/USER/.local/share
 $path = $bd->data_home('Foo', 'Bar', 'Baz');
 # returns: /home/USER/.local/share/Foo/Bar/Baz

=head2 data_dirs

 # :lookup
 my $dir = data_dirs(@path);
 my $dir = $bd->data_dirs(@path);
 my @dirs = data_dirs(@path);
 my @dirs = $bd->data_dirs(@path);

Looks for directories specified by C<@path> in the data home and
other data directories. Returns (possibly empty) list of readable
directories. In scalar context only the first directory found is
returned. Use this to lookup application data.

=head2 data_files

 # :lookup
 my $file = data_files(@path);
 my $file = $bd->data_files(@path);
 my @files = data_files(@path);
 my @files = $bd->data_files(@path);

Looks for files specified by C<@path> in the data home and other data
directories. Only returns files that are readable. In scalar context only
the first file found is returned. Use this to lookup application data.

=head2 config_home

 # :lookup
 my $dir = config_home(@path);
 my $dir = $bd->config_home(@path);

Takes a list of path elements and appends them to the config home
directory returning a new path. The new path does not need to exist.
Use this when writing user specific configuration.

=head2 config_dirs

 # :lookup
 my $dir = config_dirs(@path);
 my $dir = $bd->config_dirs(@path);
 my @dirs = config_dirs(@path);
 my @dirs = $bd->config_dirs(@path);

Looks for directories specified by C<@path> in the config home and
other config directories. Returns (possibly empty) list of readable
directories. In scalar context only the first directory found is
returned. Use this to lookup configuration.

=head2 config_files

 # :lookup
 my $file = config_files(@path);
 my $file = $bd->config_files(@path);
 my @files = config_files(@path);
 my @files = $bd->config_files(@path);

Looks for files specified by C<@path> in the config home and other
config directories. Returns a (possibly empty) list of files that
are readable. In scalar context only the first file found is returned.
Use this to lookup configuration.

=head2 cache_home

 # :lookup
 my $dir = cache_home(@path);
 my $dir = $bd->cache_home(@path);

Takes a list of path elements and appends them to the cache home
directory returning a new path. The new path does not need to exist.

=head2 xdg_data_home

 # :var
 my $dir = xdg_data_home;
 my $dir = $bd->xdg_data_home;

Returns either C<$ENV{XDG_DATA_HOME}> or it's default value.
Default is F<$HOME/.local/share>.

=head2 xdg_data_dirs

 # :var
 my @dirs = xdg_data_dirs;
 my @dirs = $bd->xdg_data_dirs;

Returns either C<$ENV{XDG_DATA_DIRS}> or it's default value as list.
Default is F</usr/local/share>, F</usr/share>.

=head2 xdg_config_home

 # :var
 my $dir = xdg_config_home;
 my $dir = $bd->xdg_config_home;

Returns either C<$ENV{XDG_CONFIG_HOME}> or it's default value.
Default is F<$HOME/.config>.

=head2 xdg_config_dirs

 # :var
 my @dirs = xdg_config_dirs;
 my @dirs = $bd->xdg_config_dirs;

Returns either C<$ENV{XDG_CONFIG_DIRS}> or it's default value as list.
Default is F</etc/xdg>.

=head2 xdg_cache_home

 # :var
 my $dir = xdg_cache_home;
 my $dir = $bd->xdg_cache_home;

Returns either C<$ENV{XDG_CACHE_HOME}> or it's default value.
Default is F<$HOME/.cache>.

=head1 NON-UNIX PLATFORMS

The use of L<File::Spec> ensures that all paths are returned in their native
formats regardless of platform.  On Windows this module will use the native
environment variables, rather than the default on UNIX (which is traditionally
C<$HOME>).

Please note that the specification is targeting Unix platforms only and
will only have limited relevance on other platforms. Any platform dependent
behavior in this module should be considered an extension of the spec.

=head1 BACKWARDS COMPATIBILITY

The methods C<xdg_data_files()> and C<xdg_config_files()> are exported for
backwards compatibility with version 0.02. They are identical to C<data_files()>
and C<config_files()> respectively but without the C<wantarray> behavior.

=cut
